main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala

```diff
@@ -300,6 +301,73 @@ object BackendObjType {
       xReturn(String.toTpe)
     }))
 
+  }
+  case class Struct(elms: List[BackendType]) extends BackendObjType with Generatable {
+
+    def genByteCode()(implicit flix: Flix): Array[Byte] = {
+      val cm = ClassMaker.mkClass(this.jvmName, IsFinal)
+
+      elms.indices.foreach(i => cm.mkField(IndexField(i)))
+      cm.mkConstructor(Constructor)
+      cm.mkMethod(ToStringMethod)
+
+      cm.closeClassMaker()
+    }
+
+    def IndexField(i: Int): InstanceField = InstanceField(this.jvmName, IsPublic, NotFinal, NotVolatile, s"field$i", elms(i))
+
+    def Constructor: ConstructorMethod = ConstructorMethod(this.jvmName, IsPublic, elms, Some(_ => {
+      withNames(1, elms){ case (_, variables) =>
+        thisLoad() ~
+          // super()
+          DUP() ~ INVOKESPECIAL(JavaObject.Constructor) ~
+          // this.field$i = var$j
+          // fields are numbered consecutively while variables skip indices based
+          // on their stack size
+          composeN(variables.zipWithIndex.map{case (elm, i) =>
+            DUP() ~ elm.load() ~ PUTFIELD(IndexField(i))
+          }) ~
+          RETURN()
+      }
+    }))
+
+    def ToStringMethod: InstanceMethod = JavaObject.ToStringMethod.implementation(this.jvmName, Some(_ => {
+      // [...] -> [..., "(v1, v2, ...)"]
+      def commaSepElmString(): InstructionSet = {
+        // new String[elms.length] // referred to as `elms`
+        cheat(mv => GenExpression.compileInt(elms.length)(mv)) ~ ANEWARRAY(String.jvmName) ~
+          ICONST_M1() ~ // running index referred to as `j`
+          // current stack [elms, j]
+          composeN(elms.indices.map(i => {
+            val field = IndexField(i)
+            // j = j + 1
+            ICONST_1() ~ IADD() ~
+              // [elms, j] -> [elms, j, elms, j]
+              DUP2() ~
+              // this.field$i.toString
+              thisLoad() ~ GETFIELD(field) ~ xToString(field.tpe) ~
+              // [elms, j, elms, j, string] -> [elms, j]
+              AASTORE()
+          })) ~
+          POP() ~
+          // [elms] -> [", ", elms]
+          pushString(", ") ~ SWAP() ~
+          INVOKESTATIC(String.JoinMethod)
+      }
+      // new String[3] // referred to as `arr`
+      ICONST_3() ~ ANEWARRAY(String.jvmName) ~
+        // arr[0] = "("
+        DUP() ~ ICONST_0() ~ pushString("(") ~ AASTORE() ~
```

Comment:
I think that instead of returning the string `(v1, v2, ...)` it should be `Struct(v1, v2, ...)`. So just change `pushString("(")` to `pushString("Struct(")`